### ISSUE TITLE
running Ansible playbook after bootstrapping Prefix fails on RHEL8/aarch64 AWS instance

### ISSUE THREAD

**Issue by boegel** [EESSI/compatibility-layer#56]
After bootstrapping Prefix via the container image (`singularity run --bind $HOME/cvmfs:/cvmfs docker://eessi/bootstrap-prefix:centos8-$(uname -m)`), I tried running the Ansible playbook to take care of the remaining tasks to set up a new version of the compat layer.

Here's what I did:

* symlink `/cvmfs` to `$HOME/cvmfs` on the build host where I performed the Prefix bootstrap
* run the Ansible playbook from another host, as follows:
  * ensure passwordless `ssh` to build host works (by adding SSH key to SSH agent via `ssh-add`)
  * create `hosts` file:
    ```
    $ cd compatibility-layer/ansible/playbooks
    $ cat hosts
    [cvmfsstratum0servers]
    <IP_OF_BUILD_HOST>
    ```
    * run playbook:
    ```
    ansible-playbook -i hosts -K install.yml -e gentoo_prefix_path=/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64
    ```

The playbook failed when running `emerge gentoolkit` (see details below).
<details>

```wrap
TASK [compatibility_layer : Install equery command (dependency for the portage module)] ***************************************************
fatal: [18.223.116.164]: FAILED! => {"changed": true, "cmd": ["emerge", "gentoolkit"], "delta": "0:00:01.930018", "end": "2020-11-03 18:41:41.190089", "msg": "non-zero return code", "rc": 1, "start": "2020-11-03 18:41:39.260071", "stderr": "portage: 'portage' user or group missing.\n         For the defaults, line 1 goes into passwd, and 2 into group.\n         portage:x:250:250:portage:/var/tmp/portage:/bin/false\n         portage::250:portage\n*** WARNING ***  For security reasons, only system administrators should be\n*** WARNING ***  allowed in the portage group.  Untrusted users or processes\n*** WARNING ***  can potentially exploit the portage group for attacks such as\n*** WARNING ***  local privilege escalation.\n\nmount: /proc: must be superuser to use mount.\nUnable to mark /proc slave: 32", "stderr_lines": ["portage: 'portage' user or group missing.", "         For the defaults, line 1 goes into passwd, and 2 into group.", "         portage:x:250:250:portage:/var/tmp/portage:/bin/false", "         portage::250:portage", "*** WARNING ***  For security reasons, only system administrators should be", "*** WARNING ***  allowed in the portage group.  Untrusted users or processes", "*** WARNING ***  can potentially exploit the portage group for attacks such as", "*** WARNING ***  local privilege escalation.", "", "mount: /proc: must be superuser to use mount.", "Unable to mark /proc slave: 32"], "stdout": "\nPerforming Global Updates\n(Could take a couple of minutes if you have a lot of binary packages.)\n  .='update pass'  *='binary update'  #='/var/db update'  @='/var/db move'\n  s='/var/db SLOT move'  %='binary move'  S='binary SLOT move'\n  p='update /etc/portage/package.*'\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2015.....................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2015..........................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/1Q-2016............................................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/2Q-2016.........................................................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2016.......................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2016.............................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/1Q-2017..............................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/2Q-2017....\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2017........................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2017.......\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/1Q-2018.....................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/2Q-2018..................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2018...\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2018.......\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/1Q-2019.................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/2Q-2019...\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2019.......\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2019.......\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/1Q-2020............\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/2Q-2020............\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2020....................................................................................................................................................\n/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2020...\n\n\nCalculating dependencies  \n * IMPORTANT: 4 news items need reading for repository 'gentoo'.\n * Use eselect news read to view new items.\n\n... done!\n\n>>> Verifying ebuild manifests\n\n>>> Emerging (1 of 1) app-portage/gentoolkit-0.5.0-r2::gentoo\n * The ebuild phase 'die_hooks' has been aborted since PORTAGE_BUILDDIR\n * does not exist: '/home/ec2-user/cvmfs/pilot.eessi-\n * hpc.org/2020.10/compat/aarch64/var/tmp/portage/app-\n * portage/gentoolkit-0.5.0-r2'\n\n>>> Failed to emerge app-portage/gentoolkit-0.5.0-r2\n * Messages for package app-portage/gentoolkit-0.5.0-r2:", "stdout_lines": ["", "Performing Global Updates", "(Could take a couple of minutes if you have a lot of binary packages.)", "  .='update pass'  *='binary update'  #='/var/db update'  @='/var/db move'", "  s='/var/db SLOT move'  %='binary move'  S='binary SLOT move'", "  p='update /etc/portage/package.*'", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2015.....................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2015..........................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/1Q-2016............................................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/2Q-2016.........................................................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2016.......................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2016.............................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/1Q-2017..............................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/2Q-2017....", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2017........................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2017.......", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/1Q-2018.....................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/2Q-2018..................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2018...", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2018.......", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/1Q-2019.................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/2Q-2019...", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2019.......", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2019.......", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/1Q-2020............", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/2Q-2020............", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/3Q-2020....................................................................................................................................................", "/home/ec2-user/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/db/repos/gentoo/profiles/updates/4Q-2020...", "", "", "Calculating dependencies  ", " * IMPORTANT: 4 news items need reading for repository 'gentoo'.", " * Use eselect news read to view new items.", "", "... done!", "", ">>> Verifying ebuild manifests", "", ">>> Emerging (1 of 1) app-portage/gentoolkit-0.5.0-r2::gentoo", " * The ebuild phase 'die_hooks' has been aborted since PORTAGE_BUILDDIR", " * does not exist: '/home/ec2-user/cvmfs/pilot.eessi-", " * hpc.org/2020.10/compat/aarch64/var/tmp/portage/app-", " * portage/gentoolkit-0.5.0-r2'", "", ">>> Failed to emerge app-portage/gentoolkit-0.5.0-r2", " * Messages for package app-portage/gentoolkit-0.5.0-r2:"]}
TASK [compatibility_layer : Abort transaction] ********************************************************************************************
skipping: [18.223.116.164]
RUNNING HANDLER [compatibility_layer : Generate locales] **********************************************************************************
changed: [18.223.116.164]
PLAY RECAP ********************************************************************************************************************************
18.223.116.164             : ok=10   changed=4    unreachable=0    failed=0    skipped=3    rescued=1    ignored=0  
```
</details>

Running `emerge gentoolkit` in the `startprefix` environment manually works fine.

When I re-run the playbook after doing that, it fails when running `emerge eselect-repository` (see details below).
<details>

```
TASK [compatibility_layer : Install eselect-repository] ***********************************************************************************
fatal: [18.223.116.164]: FAILED! => {"changed": false, "cmd": ["/cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/usr/bin/emerge", "--noreplace", "--ask=n", "eselect-repository"], "msg": "Packages not installed.", "rc": 1, "stderr": "portage: 'portage' user or group missing.\n         For the defaults, line 1 goes into passwd, and 2 into group.\n         portage:x:250:250:portage:/var/tmp/portage:/bin/false\n         portage::250:portage\n*** WARNING ***  For security reasons, only system administrators should be\n*** WARNING ***  allowed in the portage group.  Untrusted users or processes\n*** WARNING ***  can potentially exploit the portage group for attacks such as\n*** WARNING ***  local privilege escalation.\n\nmount: /proc: must be superuser to use mount.\nUnable to mark /proc slave: 32\n", "stderr_lines": ["portage: 'portage' user or group missing.", "         For the defaults, line 1 goes into passwd, and 2 into group.", "         portage:x:250:250:portage:/var/tmp/portage:/bin/false", "         portage::250:portage", "*** WARNING ***  For security reasons, only system administrators should be", "*** WARNING ***  allowed in the portage group.  Untrusted users or processes", "*** WARNING ***  can potentially exploit the portage group for attacks such as", "*** WARNING ***  local privilege escalation.", "", "mount: /proc: must be superuser to use mount.", "Unable to mark /proc slave: 32"], "stdout": "Calculating dependencies  \n * IMPORTANT: 4 news items need reading for repository 'gentoo'.\n * Use eselect news read to view new items.\n\n... done!\n\n>>> Verifying ebuild manifests\n\n>>> Emerging (1 of 8) dev-python/certifi-10001-r1::gentoo\n * Fetching files in the background.\n * To view fetch progress, run in another terminal:\n * tail -f /cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/log/emerge-fetch.log\n * The ebuild phase 'die_hooks' has been aborted since PORTAGE_BUILDDIR\n * does not exist: '/home/ec2-user/cvmfs/pilot.eessi-\n * hpc.org/2020.10/compat/aarch64/var/tmp/portage/dev-\n * python/certifi-10001-r1'\n\n>>> Failed to emerge dev-python/certifi-10001-r1\n * Messages for package dev-python/certifi-10001-r1:\n\n\n", "stdout_lines": ["Calculating dependencies  ", " * IMPORTANT: 4 news items need reading for repository 'gentoo'.", " * Use eselect news read to view new items.", "", "... done!", "", ">>> Verifying ebuild manifests", "", ">>> Emerging (1 of 8) dev-python/certifi-10001-r1::gentoo", " * Fetching files in the background.", " * To view fetch progress, run in another terminal:", " * tail -f /cvmfs/pilot.eessi-hpc.org/2020.10/compat/aarch64/var/log/emerge-fetch.log", " * The ebuild phase 'die_hooks' has been aborted since PORTAGE_BUILDDIR", " * does not exist: '/home/ec2-user/cvmfs/pilot.eessi-", " * hpc.org/2020.10/compat/aarch64/var/tmp/portage/dev-", " * python/certifi-10001-r1'", "", ">>> Failed to emerge dev-python/certifi-10001-r1", " * Messages for package dev-python/certifi-10001-r1:", "", ""]}
```

</details>

Running `emerge eselect-repository` works fine, so does `emerge dev-vcs/git` (which is the next step that fails in the playbook).

Re-running the playbook after running those 3 `emerge` commands that were failing manually makes it fail on `emerge @2020.10` (which also fails manually due to another problem, see https://github.com/EESSI/gentoo-overlay/pull/24).

Ansible version info (installed via `pip3 install ansible`):

```
$ ansible --version
ansible 2.10.3
  config file = None
  configured module search path = ['/home/ec2-user/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/linuxbrew/.linuxbrew/opt/python@3.8/lib/python3.8/site-packages/ansible
  executable location = /home/linuxbrew/.linuxbrew/bin/ansible
  python version = 3.8.6 (default, Oct 10 2020, 07:54:55) [GCC 5.4.0 20160609]
```

**Comment by terjekv** [comment]
Yeah, I saw that portage user or group missing error as well when I was testing.

**Comment by bedroge** [comment]
I also see that portage user/group warning quite often, but don't think it should cause it to crash. This seems to be the real error:

```
The ebuild phase 'die_hooks' has been aborted since PORTAGE_BUILDDIR\n * does not exist: '/home/ec2-user/cvmfs/pilot.eessi-\n * hpc.org/2020.10/compat/aarch64/var/tmp/portage/dev-\n * python/certifi-10001-r1
```

But I'm not sure what's causing this. Found a similar error here:
https://blog.dob.sk/2019/03/01/fix-for-gentoo-the-ebuild-phase-die_hooks-has-been-aborted/

Maybe it's because the task is running as root? Could be worth a try to remove the `become: True` from the `install.yml` playbook. I added that a while ago because I had to run it as root on the Stratum 0, but probably this is often not the case, so I think it may be better to remove it again (and pass `-b` on the command line, if necessary).
